### PR TITLE
BUILD-920: change default installation namespace

### PR DESCRIPTION
**Pull request by avinal**
## Changes

- by default openshift-builds operator should install in its own namespace
  to enable independent upgrades from other operators
- user can still choose their own namespace while installing

Signed-off-by: Avinal Kumar <avinal@redhat.com>

### CODE REVIEW

**Comment by sayan-biswas** [comment]
Are there any other changes you are expecting in this PR? (since its in draft)

**Comment by avinal** [comment]
Yes, some code changes are required because currently the namespace is a constant in the code and doesn't covers if installed in a different namespace.

File: internal/shipwright/build/build.go [hunk 1]
```diff
@@ -68,7 +68,7 @@ func (sb *ShipwrightBuild) CreateOrUpdate(ctx context.Context, owner client.Obje
 
 	return ctrl.CreateOrUpdate(ctx, sb.Client, object, func() error {
 		object.Spec = shipwrightv1alpha1.ShipwrightBuildSpec{
-			TargetNamespace: common.OpenShiftBuildNamespaceName,
+			TargetNamespace: owner.GetNamespace(),
```
Comment (adambkaplan) [inline internal/shipwright/build/build.go:71]: IIRC the `ShipwrightBuild` object is cluster-scoped, so `GetNamespace()` will always return an empty string.

File: internal/shipwright/build/build_test.go [hunk 2]
```diff
@@ -79,7 +79,7 @@ var _ = Describe("Build", Label("shipwright", "build"), func() {
 				Expect(metav1.IsControlledBy(object, owner)).To(BeTrue())
 			})
 			It("should have target namespace set to openshift build", func() {
-				Expect(object.Spec.TargetNamespace).To(Equal(common.OpenShiftBuildNamespaceName))
+				Expect(object.Spec.TargetNamespace).To(Equal(owner.GetNamespace()))
```
Comment (adambkaplan) [inline internal/shipwright/build/build_test.go:82]: Ditto in the tests - `ShipwrightBuild` is cluster-scoped

**Review by adambkaplan — CHANGES_REQUESTED** [review]

File: internal/shipwright/build/build.go (hunk 1 above)
Comment (sayan-biswas) [reply internal/shipwright/build/build.go:71]: Yes, `ShipwrightBuild` namespace will be blank. 
You can use the downward api to populate an environment variable or more convenient you can get it from `/var/run/secrets/kubernetes.io/serviceaccount/namespace` inside the container.

File: internal/shipwright/build/build.go [hunk 3]
```diff
@@ -68,7 +68,7 @@ func (sb *ShipwrightBuild) CreateOrUpdate(ctx context.Context, owner client.Obje
 
 	return ctrl.CreateOrUpdate(ctx, sb.Client, object, func() error {
 		object.Spec = shipwrightv1alpha1.ShipwrightBuildSpec{
-			TargetNamespace: common.OpenShiftBuildNamespaceName,
+			TargetNamespace: common.FetchCurrentNamespaceName(),
```
Comment (adambkaplan) [inline internal/shipwright/build/build.go:71]: This function does a disk read in every reconcile for something that we don't expect to change in the container's lifetime. I recommend this is called once in `cmd/main.go` and passed to the reconciler as a parameter/struct variable.

File: internal/shipwright/build/build_test.go [hunk 4]
```diff
@@ -79,7 +79,7 @@ var _ = Describe("Build", Label("shipwright", "build"), func() {
 				Expect(metav1.IsControlledBy(object, owner)).To(BeTrue())
 			})
 			It("should have target namespace set to openshift build", func() {
-				Expect(object.Spec.TargetNamespace).To(Equal(common.OpenShiftBuildNamespaceName))
+				Expect(object.Spec.TargetNamespace).To(Equal(common.FetchCurrentNamespaceName()))
```
Comment (adambkaplan) [inline internal/shipwright/build/build_test.go:82]: I wouldn't invoke this function here - if this test runs within a Kubernetes container, the current namespace name could be _very_ different. In fact, if this runs within a Tekton task, the value here can change depending on how the task is configured.

Use `.To(Equal("openshift-builds"))` instead.

File: internal/shipwright/build/build_test.go [hunk 5]
```diff
@@ -110,7 +110,7 @@ var _ = Describe("Build", Label("shipwright", "build"), func() {
 				Expect(result).To(Equal(controllerutil.OperationResultUpdated))
 			})
 			It("should update the specs to match expected", func() {
-				Expect(object.Spec.TargetNamespace).To(Equal(common.OpenShiftBuildNamespaceName))
+				Expect(object.Spec.TargetNamespace).To(Equal(common.FetchCurrentNamespaceName()))
```
Comment (adambkaplan) [inline internal/shipwright/build/build_test.go:113]: Ditto - use `.To(Equal("openshift-builds"))` instead.

File: internal/common/predicate.go [hunk 6]
```diff
@@ -12,3 +14,14 @@ func IsControlledBy(object client.Object, owner *metav1.OwnerReference) bool {
 		controller.APIVersion == owner.APIVersion &&
 		controller.Kind == owner.Kind
 }
+
+// FetchCurrentNamespaceName returns namespace name by using information stored as file
+// Returns default Openshift Builds namespace on error
+// Refer: https://kubernetes.io/docs/tasks/run-application/access-api-from-pod/#without-using-a-proxy
+func FetchCurrentNamespaceName() string {
+	content, err := os.ReadFile("/var/run/secrets/kubernetes.io/serviceaccount/namespace")
```
Comment (adambkaplan) [inline internal/common/predicate.go:22]: Per the linked documentation:

```
Finally, the default namespace to be used for namespaced API operations is placed in a file at /var/run/secrets/kubernetes.io/serviceaccount/namespace in each container.
```

I _assume_ this is the namespace the pod is deployed in?

File: bundle/manifests/openshift-builds-operator.clusterserviceversion.yaml [hunk 7]
```diff
@@ -47,6 +47,7 @@ metadata:
     features.operators.openshift.io/token-auth-aws: "false"
     features.operators.openshift.io/token-auth-azure: "false"
     features.operators.openshift.io/token-auth-gcp: "false"
+    operatorframework.io/suggested-namespace: openshift-builds
```
Comment (adambkaplan) [inline bundle/manifests/openshift-builds-operator.clusterserviceversion.yaml:50]: :+1:

File: config/manifests/bases/openshift-builds-operator.clusterserviceversion.yaml [hunk 8]
```diff
@@ -17,6 +17,7 @@ metadata:
     features.operators.openshift.io/token-auth-aws: "false"
     features.operators.openshift.io/token-auth-azure: "false"
     features.operators.openshift.io/token-auth-gcp: "false"
+    operatorframework.io/suggested-namespace: openshift-builds
```
Comment (adambkaplan) [inline config/manifests/bases/openshift-builds-operator.clusterserviceversion.yaml:20]: :+1:

**Review by adambkaplan — CHANGES_REQUESTED** [review]
The namespace that we deploy the operands to should be set only once at startup, as this value should not change during the pod's lifecycle.

Controller code should receive this namespace as a struct variable/parameter, and pass it through to relevant code as function arguments/param values.

File: internal/shipwright/build/build.go (hunk 3 above)
Comment (sayan-biswas) [reply internal/shipwright/build/build.go:71]: ++1

File: internal/common/constants.go [hunk 9]
```diff
@@ -20,4 +19,5 @@ const (
 var (
 	ShipwrightBuildManifestPath         = filepath.Join("config", "shipwright", "build", "release")
 	ShipwrightBuildStrategyManifestPath = filepath.Join("config", "shipwright", "build", "strategy")
+	OpenShiftBuilNamespaceName          string
```
Comment (adambkaplan) [inline internal/common/constants.go:22]: Please keep this a constant, as this is our "fallback" value.

File: cmd/main.go [hunk 10]
```diff
@@ -52,6 +53,9 @@ func init() {
 	utilruntime.Must(operatorv1alpha1.AddToScheme(scheme))
 	utilruntime.Must(shipwrightv1alpha1.AddToScheme(scheme))
 	//+kubebuilder:scaffold:scheme
+
+	// Fetch the namespace and store for later use
+	common.OpenShiftBuilNamespaceName = common.FetchCurrentNamespaceName()
```
Comment (adambkaplan) [inline cmd/main.go:58]: Yikes! Changing a global variable that is intended to be a constant makes code super hard to follow. This will only lead to thorny, hard to debug / "works on my cluster" bugs.

I also would not invoke `FetchCurrentNamespaceName` in the `init` function - this is reserved for stuff that should work 100% of the time in production. Since this function does a disk read, an I/O failure is possible.

File: cmd/main.go [hunk 11]
```diff
@@ -113,7 +117,7 @@ func main() {
 	buildReconciler := &controller.OpenShiftBuildReconciler{
 		Client:     mgr.GetClient(),
 		Scheme:     mgr.GetScheme(),
-		Shipwright: shipwrightbuild.New(mgr.GetClient()),
+		Shipwright: shipwrightbuild.New(mgr.GetClient(), common.OpenShiftBuilNamespaceName),
```
Comment (adambkaplan) [inline cmd/main.go:120]: I'd simply pass this in here, or do something like

```suggestion
                namespace := common.FetchCurrentNamespace()
                setupLog.Info("operator namespace found", "namespace", namespace)
		Shipwright: shipwrightbuild.New(mgr.GetClient(),  namespace),
```

File: internal/common/predicate.go [hunk 12]
```diff
@@ -12,3 +14,14 @@ func IsControlledBy(object client.Object, owner *metav1.OwnerReference) bool {
...
+	if err != nil {
+		return "openshift-builds"
```
Comment (adambkaplan) [inline internal/common/predicate.go:24]: Use the constant for the fallback value.

File: internal/common/predicate.go [hunk 13]
```diff
@@ -12,3 +14,14 @@ func IsControlledBy(object client.Object, owner *metav1.OwnerReference) bool {
...
+	}
+	return string(content)
```
Comment (adambkaplan) [inline internal/common/predicate.go:26]: Do we need to worry about line endings or other invalid characters?

File: internal/shipwright/build/build.go [hunk 14]
```diff
@@ -15,13 +15,15 @@ import (
...
 // ShipwrightBuild type defines methods to Get, Create, Delete v1alpha1.ShipwrightBuild resource
 type ShipwrightBuild struct {
-	Client client.Client
+	Client    client.Client
+	Namespace string
 }
 
 // New creates new instance of ShipwrightBuild type
-func New(client client.Client) *ShipwrightBuild {
+func New(client client.Client, namespace string) *ShipwrightBuild {
```
Comment (adambkaplan) [inline internal/shipwright/build/build.go:23]: :+1: yes!

File: internal/common/predicate.go [hunk 15]
```diff
@@ -12,3 +14,14 @@ func IsControlledBy(object client.Object, owner *metav1.OwnerReference) bool {
 		controller.APIVersion == owner.APIVersion &&
 		controller.Kind == owner.Kind
 }
+
+// FetchCurrentNamespaceName returns namespace name by using information stored as file
+// Returns default Openshift Builds namespace on error
+// Refer: https://kubernetes.io/docs/tasks/run-application/access-api-from-pod/#without-using-a-proxy
+func FetchCurrentNamespaceName() string {
```
Comment (adambkaplan) [inline internal/common/predicate.go:21]: :thinking: Since this thing is always running in a Kubernetes pod with a service account + token, I wonder if we should have this return error if we can't read the file.

I think we can merge this as is and discuss once it lands.

File: internal/common/predicate.go (hunk 6 above)
Comment (adambkaplan) [inline internal/common/predicate.go:22]: An area for future improvement - create an "inner" function that captures the rest of the logic, but accepts the file name as a parameter. This makes the code unit-testable outside of Kubernetes.

**Review by adambkaplan — CHANGES_REQUESTED** [review]
Only one major "must change this" regarding the `OpenshiftBuildNamespace` constant. Everything else is nice to haves/areas of future improvement.

File: internal/common/predicate.go (hunk 13 above)
Comment (avinal) [reply internal/common/predicate.go:26]: No, because the file has only one string with no line ending, also we are doing a file read instead of line read and then passing the value to a byte to string converter so that ensure that our string is free of other chars.

File: internal/common/predicate.go (hunk 15 above)
Comment (avinal) [reply internal/common/predicate.go:21]: I think we should have a fallback logic for fetching namespace.

File: README.md [hunk 16]
```diff
@@ -67,33 +73,42 @@ make undeploy
...
+   make bundle-build bundle-push IMAGE_TAG_BASE=quay.io/myusername/rh-openshift-builds/operator
    ```
 
 3. Build and push the operator catalog
 
    ```sh
-   $ make catalog-fbc-build IMAGE_TAG_BASE=quay.io/myusername/rh-openshift-builds/operator
-   $ make catalog-push IMAGE_TAG_BASE=quay.io/myusername/rh-openshift-builds/operator
+   make catalog-fbc-build IMAGE_TAG_BASE=quay.io/myusername/rh-openshift-builds/operator
+   make catalog-push IMAGE_TAG_BASE=quay.io/myusername/rh-openshift-builds/operator
```
Comment (apoorvajagtap) [inline README.md:96]:
```suggestion
   make catalog-build catalog-push IMAGE_TAG_BASE=quay.io/myusername/rh-openshift-builds/operator
```
I think, `catalog-build` has been missed? cc @sayan-biswas

File: README.md [hunk 17]
```diff
@@ -67,33 +73,42 @@ make undeploy
 Red Hat operators are designed to be managed by Operator Lifecycle Manager (OLM) and deployed
 through the `OperatorHub` section of the OpenShift web console. To deploy with OpenShift and OLM:
 
+> [!TIP]
+> You can combine all or multiple `make` commands in one single command if you don't have anything to modify in between the steps.
+>
+> ```sh
+> make docker-build docker-push bundle bundle-build bundle-push catalog-fbc-build catalog-push catalog-deploy IMAGE_TAG_BASE=quay.io/musername/rh-openshift-builds/operator
```
Comment (apoorvajagtap) [inline README.md:80]:
```suggestion
> make docker-build docker-push bundle bundle-build bundle-push catalog-fbc-build catalog-build catalog-push catalog-deploy IMAGE_TAG_BASE=quay.io/musername/rh-openshift-builds/operator
```
same.

**Review by apoorvajagtap — CHANGES_REQUESTED** [review]

File: internal/common/constants.go [hunk 18]
```diff
@@ -20,4 +20,5 @@ const (
 var (
 	ShipwrightBuildManifestPath         = filepath.Join("config", "shipwright", "build", "release")
 	ShipwrightBuildStrategyManifestPath = filepath.Join("config", "shipwright", "build", "strategy")
+	CurrentNamespaceName                string
```
Comment (adambkaplan) [inline internal/common/constants.go:23]: I still don't fully understand why we are storing this in a global state.

File: internal/common/predicate.go [hunk 19]
```diff
@@ -12,3 +14,16 @@ func IsControlledBy(object client.Object, owner *metav1.OwnerReference) bool {
...
+	namespace, err := os.ReadFile("/var/run/secrets/kubernetes.io/serviceaccount/namespace")
+	if err != nil {
+		CurrentNamespaceName = OpenShiftBuilNamespaceName
+	} else {
+		CurrentNamespaceName = string(namespace)
+	}
+	return CurrentNamespaceName
```
Comment (adambkaplan) [inline internal/common/predicate.go:28]: Avoid global state/variables here. Suggestion also includes the typo fix.

```suggestion
	if err != nil {
		return OpenShiftBuildNamespaceName
	}
	return string(namespace)
```

**Review by adambkaplan — APPROVED** [review]
Not blocking merge on this - items that have been identified are minor things that can be addressed in a follow-up pull request.
